pkg/controllers/metrics/node/counter.go

```diff
@@ -0,0 +1,241 @@
+			Subsystem: metricSubsystem,
+			Name:      "ready_node_count",
+			Help:      "Count of nodes that are ready by provisioner and zone.",
+		},
+		[]string{
+			metricLabelProvisioner,
+			metricLabelZone,
+		},
+	)
+
+	readyNodeCountByArchProvisionerZone = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: metricNamespace,
+			Subsystem: metricSubsystem,
+			Name:      "ready_node_arch_count",
+			Help:      "Count of nodes that are ready by architecture, provisioner, and zone.",
+		},
+		[]string{
+			metricLabelArch,
+			metricLabelProvisioner,
+			metricLabelZone,
+		},
+	)
+
+	readyNodeCountByInstancetypeProvisionerZone = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: metricNamespace,
+			Subsystem: metricSubsystem,
+			Name:      "ready_node_instancetype_count",
+			Help:      "Count of nodes that are ready by instance type, provisioner, and zone.",
+		},
+		[]string{
+			metricLabelInstanceType,
+			metricLabelProvisioner,
+			metricLabelZone,
+		},
+	)
+
+	readyNodeCountByOsProvisionerZone = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: metricNamespace,
+			Subsystem: metricSubsystem,
+			Name:      "ready_node_os_count",
+			Help:      "Count of nodes that are ready by operating system, provisioner, and zone.",
+		},
+		[]string{
+			metricLabelOS,
+			metricLabelProvisioner,
+			metricLabelZone,
+		},
+	)
+)
+
+func init() {
+	crmetrics.Registry.MustRegister(nodeCountByProvisioner)
+	crmetrics.Registry.MustRegister(readyNodeCountByProvisionerZone)
+	crmetrics.Registry.MustRegister(readyNodeCountByArchProvisionerZone)
+	crmetrics.Registry.MustRegister(readyNodeCountByInstancetypeProvisionerZone)
+	crmetrics.Registry.MustRegister(readyNodeCountByOsProvisionerZone)
+}
+
+func publishNodeCountsForProvisioner(provisioner string, consumeNodesWith consumeNodesWithFunc) error {
+	archValues := knownValuesForNodeLabels[nodeLabelArch]
+	instanceTypeValues := knownValuesForNodeLabels[nodeLabelInstanceType]
+	osValues := knownValuesForNodeLabels[nodeLabelOS]
+	zoneValues := knownValuesForNodeLabels[nodeLabelZone]
+
+	errors := make([]error, 0, len(archValues)*len(instanceTypeValues)*len(osValues)*len(zoneValues))
+
+	// 1. Publish the count of all nodes associated with `provisioner`.
+	nodeLabels := client.MatchingLabels{nodeLabelProvisioner: provisioner}
+	errors = append(errors, consumeNodesWith(nodeLabels, func(nodes []v1.Node) error {
+		return publishCount(nodeCountByProvisioner, metricLabelsFrom(nodeLabels), len(nodes))
+	}))
+
+	// 2. Publish the count of all nodes associated with `provisioner`, in `zone`, and reported as "ready".
+	// 3. Publish the count of all nodes with `arch`, associated with `provisioner`, in `zone`, and reported as "ready".
+	// 4. Publish the count of all nodes with `instanceType`, associated with `provisioner`, in `zone`, and reported as "ready"
+	// 5. Publish the count of all nodes with `os`, associated with `provisioner`, in `zone`, and reported as "ready".
```

Comment:
Honestly I don't think these comments are really necessary, they pretty much just restate the code - you could just say Publish counts of nodes along different dimensions, or something to that effect (otherwise, every time a new count is added, the comment will have to be updated).